File: proxy.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

var proxyModels = []map[string]any{
	{"id": "deepseek-chat", "object": "model", "created": 1700000000, "owned_by": "deepseek"},
	{"id": "deepseek-reasoner", "object": "model", "created": 1700000000, "owned_by": "deepseek"},
}

// ProxyServer is the OpenAI-compatible proxy (port of proxy.py).
type ProxyServer struct {
	log      *log.Logger
	proxyKey string

	mu     sync.Mutex
	api    *DeepSeekAPI
	apiErr error
	apiSet bool

	hl         sync.Mutex
	histChatID string
	histParID  any
	useHistory bool
}

func NewProxyServer(logger *log.Logger, proxyKey string) *ProxyServer {
	return &ProxyServer{log: logger, proxyKey: proxyKey}
}

func (s *ProxyServer) getAPI() (*DeepSeekAPI, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.apiSet {
		return s.api, s.apiErr
	}
	token := os.Getenv("DEEPSEEK_TOKEN")
	if token == "" {
		s.apiErr = errors.New("DEEPSEEK_TOKEN not set.\n  export DEEPSEEK_TOKEN=<chat.deepseek.com localStorage -> userToken -> value>")
	} else {
		s.api, s.apiErr = NewDeepSeekAPI(token)
		if s.apiErr == nil {
			s.log.Printf("DeepSeekAPI ready")
		}
	}
	s.apiSet = true
	return s.api, s.apiErr
}

func (s *ProxyServer) newHist() (string, error) {
	api, err := s.getAPI()
	if err != nil {
		return "", err
	}
	cid, err := api.CreateChatSession(context.Background())
	if err != nil {
		return "", err
	}
	s.hl.Lock()
	s.histChatID, s.histParID = cid, nil
	s.hl.Unlock()
	s.log.Printf("New history session: %s", cid)
	return cid, nil
}

func (s *ProxyServer) getHist() (string, any) {
	s.hl.Lock()
	defer s.hl.Unlock()
	return s.histChatID, s.histParID
}

func (s *ProxyServer) setHistPar(mid any) {
	s.hl.Lock()
	defer s.hl.Unlock()
	s.histParID = mid
}

func (s *ProxyServer) setUseHistory(v bool) {
	s.hl.Lock()
	defer s.hl.Unlock()
	s.useHistory = v
}

func (s *ProxyServer) getUseHistory() bool {
	s.hl.Lock()
	defer s.hl.Unlock()
	return s.useHistory
}

func (s *ProxyServer) checkAuth(r *http.Request) bool {
	if s.proxyKey == "" {
		return true
	}
	return r.Header.Get("Authorization") == "Bearer "+s.proxyKey
}

func (s *ProxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	cors := func() {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")
	}

	if r.Method == http.MethodOptions {
		cors()
		w.WriteHeader(200)
		return
	}
	cors()

	if debugMode {
		debugDumpClientRequest(r, nil) // headers only; handlers dump the body
	}

	if !s.checkAuth(r) {
		writeError(w, "Bad token", "authentication_error", "invalid_api_key", 401)
		return
	}

	path := r.URL.Path
	if r.Method == http.MethodGet {
		switch path {
		case "/v1/models", "/models":
			writeJSON(w, http.StatusOK, map[string]any{"object": "list", "data": proxyModels})
			return
		case "/history":
			q := r.URL.Query()
			v := q.Get("enable")
			if v == "" {
				v = q.Get("value")
			}
			enabled := strings.EqualFold(v, "true")
			s.setUseHistory(enabled)
			cid, _ := s.getHist()
			writeJSON(w, http.StatusOK, map[string]any{"use_history": enabled, "chat_id": cid})
			return
		}
		writeError(w, "Not Found", "invalid_request_error", "not_found", 404)
		return
	}

	if r.Method == http.MethodPost {
		switch path {
		case "/new":
			if _, err := s.getAPI(); err != nil {
				writeError(w, err.Error(), "authentication_error", "missing_token", 401)
				return
			}
			cid, err := s.newHist()
			if err != nil {
				writeError(w, err.Error(), "upstream_error", nil, 502)
				return
			}
			writeJSON(w, http.StatusOK, map[string]any{"message": "New session", "chat_id": cid})
			return
		case "/history":
			var body map[string]any
			if err := decodeBody(w, r, &body); err != nil {
				return
			}
			var enabled bool
			if v, ok := body["enable"]; ok {
				enabled = toBool(v)
			} else if v, ok := body["value"]; ok {
				enabled = toBool(v)
			}
			s.setUseHistory(enabled)
			writeJSON(w, http.StatusOK, map[string]any{"use_history": enabled})
			return
		case "/v1/chat/completions", "/chat/completions":
			s.handleChat(w, r)
			return
		}
		writeError(w, "Not Found", "invalid_request_error", "not_found", 404)
		return
	}

	writeError(w, "Not Found", "invalid_request_error", "not_found", 404)
}

// chatMessage is one OpenAI-style message. Content stays raw so both string
// and typed-part arrays are accepted.
type chatMessage struct {
	Role       string              `json:"role"`
	Content    json.RawMessage     `json:"content"`
	ToolCallID string              `json:"tool_call_id,omitempty"`
	ToolCalls  []assistantToolCall `json:"tool_calls,omitempty"`
	Name       string              `json:"name,omitempty"`
}

// chatRequest mirrors the OpenAI-style request body.
type chatRequest struct {
	Messages  []chatMessage `json:"messages"`
	Model     string        `json:"model"`
	Stream    bool          `json:"stream"`
	Thinking  bool          `json:"thinking"`
	DeepThink bool          `json:"deepThink"`
	Search    bool          `json:"search"`
	Tools     []openAITool  `json:"tools,omitempty"`
}

func (s *ProxyServer) handleChat(w http.ResponseWriter, r *http.Request) {
	var body chatRequest
	rawBody, err := decodeBodyRaw(w, r, &body)
	if err != nil {
		return
	}
	if debugMode {
		debugDumpClientRequest(r, rawBody)
	}

	model := body.Model
	if model == "" {
		model = "deepseek-chat"
	}
	stream := body.Stream
	thinking := body.Thinking || body.DeepThink || model == "deepseek-reasoner"
	search := body.Search

	prompt := ""
	if agentMode {
		// Agent mode folds every message plus the tool contract into one
		// role-aware prompt so the model cannot silently use tools outside
		// the caller's OpenAI tool contract (web search stays off).
		prompt = buildAgentPrompt(body.Messages, body.Tools)
		search = false
		debugf("agent prompt (%d chars):\n%s", len(prompt), prompt)
	} else if len(body.Messages) > 0 {
		raw := body.Messages[len(body.Messages)-1].Content
		var s string
		if json.Unmarshal(raw, &s) == nil {
			prompt = s
		} else {
			var parts []struct {
				Type string `json:"type"`
				Text string `json:"text"`
			}
			if json.Unmarshal(raw, &parts) == nil {
				texts := []string{}
				for _, p := range parts {
					if p.Type == "text" {
						texts = append(texts, p.Text)
					}
				}
				prompt = strings.Join(texts, "\n")
			}
		}
	}
	if prompt == "" {
		prompt = " "
	}

	api, err := s.getAPI()
	if err != nil {
		writeError(w, err.Error(), "authentication_error", "missing_token", 401)
		return
	}

	var chatID string
	var parID any
	if s.getUseHistory() {
		chatID, parID = s.getHist()
		if chatID == "" {
			chatID, err = s.newHist()
			parID = nil
			if err != nil {
				writeError(w, err.Error(), "upstream_error", nil, 502)
				return
			}
		}
	} else {
		chatID, err = api.CreateChatSession(context.Background())
		parID = nil
		if err != nil {
			writeError(w, err.Error(), "upstream_error", nil, 502)
			return
		}
		s.log.Printf("Stateless session: %s", chatID)
	}

	s.log.Printf("-> model=%-20s think=%-5v search=%-5v history=%-5v agent=%v", model, thinking, search, s.getUseHistory(), agentMode)

	params := ChatParams{
		ChatSessionID:   chatID,
		Prompt:          prompt,
		ParentMessageID: parID,
		ThinkingEnabled: thinking,
		SearchEnabled:   search,
	}

	if stream {
		s.streamResponse(w, r, api, params, model)
	} else {
		s.blockResponse(w, r, api, params, model)
	}
}

func (s *ProxyServer) streamResponse(w http.ResponseWriter, r *http.Request, api *DeepSeekAPI, params ChatParams, model string) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		writeError(w, "Streaming unsupported", "server_error", nil, 500)
		return
	}
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "close")
	if debugMode {
		debugDumpClientResponse(200, w.Header())
	}
	w.WriteHeader(200)

	rid := "chatcmpl-" + randomHex(16)
	created := time.Now().Unix()

	sse := func(obj map[string]any) bool {
		raw, err := json.Marshal(obj)
		if err != nil {
			return false
		}
		debugf("sse -> %s", raw)
		if _, err := w.Write(append([]byte("data: "), append(raw, []byte("\n\n")...)...)); err != nil {
			return false
		}
		flusher.Flush()
		return true
	}

	contentChunk := func(text string, finish any) map[string]any {
		return map[string]any{
			"id": rid, "object": "chat.completion.chunk", "created": created, "model": model,
			"choices": []any{map[string]any{"index": 0, "delta": map[string]any{"content": text}, "finish_reason": finish}},
		}
	}

	// role-establishing chunk (required by most clients)
	sse(contentChunkRole(rid, created, model))

	// agent mode: incrementally split model text from tool-call blocks
	var interceptor *agentStreamInterceptor
	if agentMode {
		interceptor = &agentStreamInterceptor{}
	}
	emittedToolCall := false

	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()

	err := api.ChatCompletion(ctx, params, func(ch Chunk) bool {
		if ch.Type == "ready" && s.getUseHistory() && ch.ResponseMessageID != nil {
			s.setHistPar(ch.ResponseMessageID)
		}
		if ch.Type != "content" || ch.Content == "" {
			return false
		}
		if interceptor != nil {
			parsed := interceptor.feed(ch.Content)
			if parsed.content != "" && !sse(contentChunk(parsed.content, nil)) {
				return true
			}
			for _, call := range parsed.toolCalls {
				emittedToolCall = true
				debugf("agent tool call #%d: %s(%s)", call["index"], call["function"].(map[string]any)["name"], call["function"].(map[string]any)["arguments"])
				if !sse(map[string]any{
					"id": rid, "object": "chat.completion.chunk", "created": created, "model": model,
					"choices": []any{map[string]any{"index": 0, "delta": map[string]any{"tool_calls": []any{call}}, "finish_reason": nil}},
				}) {
					return true
				}
			}
			return false
		}
		return !sse(contentChunk(ch.Content, nil))
	})

	if err != nil {
		etype, _ := errorType(err)
		sse(map[string]any{"error": map[string]any{"message": err.Error(), "type": etype}})
		return
	}

	if interceptor != nil {
		if trailing := interceptor.flush(); trailing != "" {
			sse(contentChunk(trailing, nil))
		}
	}

	finishReason := "stop"
	if emittedToolCall {
		finishReason = "tool_calls"
	}
	sse(map[string]any{
		"id": rid, "object": "chat.completion.chunk", "created": created, "model": model,
		"choices": []any{map[string]any{"index": 0, "delta": map[string]any{}, "finish_reason": finishReason}},
	})
	w.Write([]byte("data: [DONE]\n\n"))
	flusher.Flush()
}

// contentChunkRole is the role-establishing first SSE chunk (required by most
// clients).
func contentChunkRole(rid string, created int64, model string) map[string]any {
	return map[string]any{
		"id": rid, "object": "chat.completion.chunk", "created": created, "model": model,
		"choices": []any{map[string]any{"index": 0, "delta": map[string]any{"role": "assistant", "content": ""}, "finish_reason": nil}},
	}
}

func (s *ProxyServer) blockResponse(w http.ResponseWriter, r *http.Request, api *DeepSeekAPI, params ChatParams, model string) {
	ctx := r.Context()
	var parts []string

	err := api.ChatCompletion(ctx, params, func(ch Chunk) bool {
		if ch.Type == "ready" && s.getUseHistory() && ch.ResponseMessageID != nil {
			s.setHistPar(ch.ResponseMessageID)
		}
		if ch.Type == "content" && ch.Content != "" {
			parts = append(parts, ch.Content)
		}
		return false
	})

	if err != nil {
		etype, status := errorType(err)
		writeError(w, err.Error(), etype, nil, status)
		return
	}

	answer := strings.Join(parts, "")
	finishReason := "stop"
	message := map[string]any{"role": "assistant", "content": answer}

	if agentMode {
		if calls := parseAgentToolCalls(answer); len(calls) > 0 {
			finishReason = "tool_calls"
			message["tool_calls"] = calls
			debugf("agent tool calls parsed: %d", len(calls))
			debugf("agent tool calls: %s", debugJSON(calls))
		}
		answer = stripAgentToolCalls(answer)
		message["content"] = answer
	}

	promptTokens := countWords(params.Prompt)
	completionTokens := countWords(answer)

	writeJSON(w, http.StatusOK, map[string]any{
		"id":      "chatcmpl-" + randomHex(16),
		"object":  "chat.completion",
		"created": time.Now().Unix(),
		"model":   model,
		"choices": []any{map[string]any{
			"index":         0,
			"message":       message,
			"finish_reason": finishReason,
		}},
		"usage": map[string]any{
			"prompt_tokens":     promptTokens,
			"completion_tokens": completionTokens,
			"total_tokens":      promptTokens + completionTokens,
		},
	})
}

// ── helpers ──────────────────────────────────────────────────────────────────

func writeJSON(w http.ResponseWriter, status int, data any) {
	raw, err := json.Marshal(data)
	if err != nil {
		writeError(w, "Internal error", "server_error", nil, 500)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	if debugMode {
		debugDumpClientResponse(status, w.Header())
		debugf("  -> body: %s", debugIndent(debugTruncate(raw)))
	}
	w.WriteHeader(status)
	w.Write(raw)
}

func writeError(w http.ResponseWriter, msg, etype string, code any, status int) {
	writeJSON(w, status, map[string]any{
		"error": map[string]any{
			"message": msg,
			"type":    etype,
			"param":   nil,
			"code":    code,
		},
	})
}

func decodeBody(w http.ResponseWriter, r *http.Request, dst any) error {
	_, err := decodeBodyRaw(w, r, dst)
	return err
}

// decodeBodyRaw decodes the JSON body into dst and also returns the raw bytes
// so debug mode can dump exactly what the client sent.
func decodeBodyRaw(w http.ResponseWriter, r *http.Request, dst any) ([]byte, error) {
	r.Body = http.MaxBytesReader(w, r.Body, 16<<20)
	raw, err := io.ReadAll(r.Body)
	if err != nil {
		writeError(w, fmt.Sprintf("Invalid JSON body: %v", err), "invalid_request_error", "invalid_json", 400)
		return nil, err
	}
	if err := json.Unmarshal(raw, dst); err != nil {
		writeError(w, fmt.Sprintf("Invalid JSON body: %v", err), "invalid_request_error", "invalid_json", 400)
		return nil, err
	}
	return raw, nil
}

func toBool(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		b, _ := strconv.ParseBool(t)
		return b
	}
	return false
}

func countWords(s string) int {
	return len(strings.Fields(s))
}

File: proxy.py
#!/usr/bin/env python3
"""
DeepSeek OpenAI-Compatible Proxy  (v6 — source verified from working api.py)
=============================================================================
Chunk types yielded by the patched api.py chat_completion():
  {'type': 'ready',          'request_message_id': N, 'response_message_id': N}
  {'type': 'session_update', 'updated_at': float}
  {'type': 'content',        'content': str, 'full_content': str}   ← TEXT
  {'type': 'status',         'status': 'FINISHED'}
  {'type': 'complete',       'finish_reason': 'stop'}
  {'type': 'title',          'content': str}

Setup:
  export DEEPSEEK_TOKEN="eyJ..."
  export PROXY_API_KEY="deepseek-proxy"
  python proxy.py
"""

import os, sys, json, time, uuid, threading, logging
from http.server import HTTPServer, BaseHTTPRequestHandler
from socketserver import ThreadingMixIn
from urllib.parse import urlparse, parse_qs

# ── import api.py (same directory) ───────────────────────────────────────────
try:
    from api import (
        DeepSeekAPI,
        AuthenticationError, RateLimitError,
        NetworkError, CloudflareError, APIError,
    )
except ImportError:
    print("[ERROR] Cannot import api.py — place proxy.py in the same folder.", file=sys.stderr)
    sys.exit(1)

# ── config ────────────────────────────────────────────────────────────────────
PORT           = int(os.getenv("PORT",          3000))
PROXY_API_KEY  = os.getenv("PROXY_API_KEY",  "deepseek-proxy")
DEEPSEEK_TOKEN = os.getenv("DEEPSEEK_TOKEN", "")

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  %(levelname)-8s  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("ds-proxy")

MODELS = [
    {"id": "deepseek-chat",     "object": "model", "created": 1700000000, "owned_by": "deepseek"},
    {"id": "deepseek-reasoner", "object": "model", "created": 1700000000, "owned_by": "deepseek"},
]

# ── singleton API ─────────────────────────────────────────────────────────────
_api_lock = threading.Lock()
_api = None

def get_api():
    global _api
    if _api: return _api
    with _api_lock:
        if not _api:
            if not DEEPSEEK_TOKEN:
                raise RuntimeError(
                    "DEEPSEEK_TOKEN not set.\n"
                    "  export DEEPSEEK_TOKEN=<chat.deepseek.com localStorage -> userToken -> value>"
                )
            _api = DeepSeekAPI(DEEPSEEK_TOKEN)
            log.info("DeepSeekAPI ready")
    return _api

# ── history state ─────────────────────────────────────────────────────────────
_hl           = threading.Lock()
_hist_chat_id = None
_hist_par_id  = None   # response_message_id from last DS response
_use_history  = False

def _new_hist():
    global _hist_chat_id, _hist_par_id
    cid = get_api().create_chat_session()
    with _hl: _hist_chat_id, _hist_par_id = cid, None
    log.info("New history session: %s", cid)
    return cid

def _get_hist():
    with _hl: return _hist_chat_id, _hist_par_id

def _set_hist_par(mid):
    global _hist_par_id
    with _hl: _hist_par_id = mid

# ── threaded server ───────────────────────────────────────────────────────────
class ThreadingHTTPServer(ThreadingMixIn, HTTPServer):
    daemon_threads = True

# ── handler ───────────────────────────────────────────────────────────────────
class Handler(BaseHTTPRequestHandler):
    def log_message(self, fmt, *args): pass

    def _body(self):
        n = int(self.headers.get("Content-Length", 0))
        return json.loads(self.rfile.read(n)) if n else {}

    def _json(self, data, status=200):
        b = json.dumps(data).encode()
        self.send_response(status)
        self.send_header("Content-Type",   "application/json")
        self.send_header("Content-Length", str(len(b)))
        self._cors(); self.end_headers(); self.wfile.write(b)

    def _err(self, msg, etype="server_error", code=None, status=500):
        self._json({"error": {"message": msg, "type": etype,
                               "param": None, "code": code}}, status)

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin",  "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")

    def _ok(self):
        if not PROXY_API_KEY: return True
        return self.headers.get("Authorization", "") == f"Bearer {PROXY_API_KEY}"

    # verbs ────────────────────────────────────────────────────────────────────

    def do_OPTIONS(self):
        self.send_response(200); self._cors(); self.end_headers()

    def do_GET(self):
        if not self._ok():
            return self._err("Bad token", "authentication_error", "invalid_api_key", 401)
        parsed = urlparse(self.path)
        qs, path = parse_qs(parsed.query), parsed.path

        if path in ("/v1/models", "/models"):
            self._json({"object": "list", "data": MODELS})

        elif path == "/history":
            global _use_history
            v = qs.get("enable", qs.get("value", ["false"]))[0]
            _use_history = v.lower() == "true"
            cid, _ = _get_hist()
            self._json({"use_history": _use_history, "chat_id": cid})

        else:
            self._err("Not Found", "invalid_request_error", "not_found", 404)

    def do_POST(self):
        if not self._ok():
            return self._err("Bad token", "authentication_error", "invalid_api_key", 401)
        path = urlparse(self.path).path
        try:
            if path == "/new":
                try: get_api()
                except RuntimeError as e:
                    return self._err(str(e), "authentication_error", "missing_token", 401)
                self._json({"message": "New session", "chat_id": _new_hist()})

            elif path == "/history":
                body = self._body()
                global _use_history
                _use_history = bool(body.get("enable", body.get("value", False)))
                self._json({"use_history": _use_history})

            elif path in ("/v1/chat/completions", "/chat/completions"):
                self._chat()

            else:
                self._err("Not Found", "invalid_request_error", "not_found", 404)

        except Exception as exc:
            log.exception("Unhandled: %s", exc)
            try: self._err(str(exc))
            except Exception: pass

    # ── /v1/chat/completions ──────────────────────────────────────────────────

    def _chat(self):
        body     = self._body()
        messages = body.get("messages", [])
        model    = body.get("model",    "deepseek-chat")
        stream   = body.get("stream",   True)
        thinking = (
            body.get("thinking",  False) or
            body.get("deepThink", False) or
            model == "deepseek-reasoner"
        )
        search = body.get("search", False)

        # prompt = content of last message
        raw = messages[-1].get("content", "") if messages else ""
        if isinstance(raw, list):
            prompt = "\n".join(p["text"] for p in raw if p.get("type") == "text")
        else:
            prompt = raw or " "

        try:
            api = get_api()
        except RuntimeError as e:
            return self._err(str(e), "authentication_error", "missing_token", 401)

        # session
        if _use_history:
            chat_id, par_id = _get_hist()
            if not chat_id:
                try:   chat_id = _new_hist(); par_id = None
                except Exception as e:
                    return self._err(str(e), "upstream_error", None, 502)
        else:
            # stateless: fresh session every request
            try:
                chat_id = api.create_chat_session()
                par_id  = None
                log.info("Stateless session: %s", chat_id)
            except Exception as e:
                return self._err(str(e), "upstream_error", None, 502)

        log.info("→ model=%-20s think=%-5s search=%-5s history=%s",
                 model, thinking, search, _use_history)

        if stream:
            self._stream(api, chat_id, par_id, prompt, model, thinking, search)
        else:
            self._block(api, chat_id, par_id, prompt, model, thinking, search)

    # ── streaming ─────────────────────────────────────────────────────────────

    def _stream(self, api, chat_id, par_id, prompt, model, thinking, search):
        self.close_connection = True  # drop TCP after stream ends
        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control","no-cache")
        self.send_header("Connection",   "close")
        self._cors(); self.end_headers()

        rid     = "chatcmpl-" + uuid.uuid4().hex
        created = int(time.time())

        def sse(obj):
            try:
                self.wfile.write(f"data: {json.dumps(obj)}\n\n".encode())
                self.wfile.flush()
            except (BrokenPipeError, ConnectionResetError): pass

        # role-establishing chunk (required by most clients)
        sse({"id": rid, "object": "chat.completion.chunk", "created": created,
             "model": model, "choices": [{"index": 0,
             "delta": {"role": "assistant", "content": ""}, "finish_reason": None}]})

        params = {
            "chat_session_id":  chat_id,
            "prompt":           prompt,
            "thinking_enabled": thinking,
            "search_enabled":   search,
        }
        if par_id is not None:
            params["parent_message_id"] = par_id

        try:
            for chunk in api.chat_completion(**params):
                ctype = chunk.get("type", "")

                # ── capture response_message_id for history threading ─────────
                if ctype == "ready" and _use_history:
                    mid = chunk.get("response_message_id")
                    if mid: _set_hist_par(mid)

                # ── only 'content' chunks carry text to forward ───────────────
                if ctype != "content":
                    continue

                text = chunk.get("content", "")
                if not text:
                    continue

                sse({"id": rid, "object": "chat.completion.chunk",
                     "created": created, "model": model,
                     "choices": [{"index": 0,
                                  "delta": {"content": text},
                                  "finish_reason": None}]})

        except AuthenticationError as e:
            sse({"error": {"message": str(e), "type": "authentication_error"}}); return
        except RateLimitError as e:
            sse({"error": {"message": str(e), "type": "rate_limit_error"}}); return
        except CloudflareError as e:
            sse({"error": {"message": str(e), "type": "cloudflare_error"}}); return
        except (NetworkError, APIError) as e:
            sse({"error": {"message": str(e), "type": "api_error"}}); return
        except (BrokenPipeError, ConnectionResetError): return
        except Exception as e:
            log.exception("Stream error: %s", e); return

        # stop chunk
        sse({"id": rid, "object": "chat.completion.chunk", "created": created,
             "model": model, "choices": [{"index": 0, "delta": {},
             "finish_reason": "stop"}]})
        try:
            self.wfile.write(b"data: [DONE]\n\n"); self.wfile.flush()
        except (BrokenPipeError, ConnectionResetError, OSError): pass
        finally:
            try: self.wfile.close()
            except OSError: pass
        self.close_connection = True

    # ── non-streaming ─────────────────────────────────────────────────────────

    def _block(self, api, chat_id, par_id, prompt, model, thinking, search):
        parts = []

        params = {
            "chat_session_id":  chat_id,
            "prompt":           prompt,
            "thinking_enabled": thinking,
            "search_enabled":   search,
        }
        if par_id is not None:
            params["parent_message_id"] = par_id

        try:
            for chunk in api.chat_completion(**params):
                ctype = chunk.get("type", "")

                if ctype == "ready" and _use_history:
                    mid = chunk.get("response_message_id")
                    if mid: _set_hist_par(mid)

                if ctype == "content":
                    text = chunk.get("content", "")
                    if text: parts.append(text)

        except AuthenticationError as e: return self._err(str(e),"authentication_error",None,401)
        except RateLimitError      as e: return self._err(str(e),"rate_limit_error",None,429)
        except CloudflareError     as e: return self._err(str(e),"cloudflare_error",None,503)
        except (NetworkError,APIError) as e: return self._err(str(e),"api_error",None,502)

        answer = "".join(parts)
        pt = len(prompt.split()); ct = len(answer.split())
        self._json({
            "id":      "chatcmpl-" + uuid.uuid4().hex,
            "object":  "chat.completion",
            "created": int(time.time()),
            "model":   model,
            "choices": [{"index": 0,
                         "message": {"role": "assistant", "content": answer},
                         "finish_reason": "stop"}],
            "usage":   {"prompt_tokens": pt, "completion_tokens": ct,
                        "total_tokens": pt + ct},
        })

# ── main ──────────────────────────────────────────────────────────────────────

def main():
    log.info("DeepSeek OpenAI Proxy  v6")
    log.info("  Port      : %d", PORT)
    log.info("  Proxy key : %s", PROXY_API_KEY or "(open)")
    log.info("  DS token  : %s", "SET" if DEEPSEEK_TOKEN else "NOT SET  ← export DEEPSEEK_TOKEN=...")
    log.info("  Endpoints : POST /v1/chat/completions")
    log.info("              GET  /v1/models")
    log.info("              GET  /history?enable=true|false")
    log.info("              POST /new")
    server = ThreadingHTTPServer(("0.0.0.0", PORT), Handler)
    log.info("Ready → http://0.0.0.0:%d/v1", PORT)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        log.info("Bye"); server.server_close()

if __name__ == "__main__":
    main()
